app/plugin.go

```diff
@@ -662,16 +674,54 @@ func (a *App) installPrepackagedPlugin(pluginPaths *pluginSignaturePaths) (*plug
 	}
 	defer os.RemoveAll(tmpDir)
 
-	plug, pluginDir, err := getPrepackagedPlugin(pluginPaths, fileReader, tmpDir)
+	plugin, pluginDir, err := getPrepackagedPlugin(pluginPaths, fileReader, tmpDir)
 	if err != nil {
 		return nil, errors.Wrapf(err, "Failed to get prepackaged plugin %s", pluginPaths.path)
 	}
 
-	if _, err := a.installExtractedPlugin(plug.Manifest, pluginDir, true); err != nil {
-		return nil, errors.Wrapf(err, "Failed to install extracted prepackaged plugin %s", pluginPaths.path)
+	if !*a.Config().PluginSettings.AutomaticPrepackagedPlugins {
+		return plugin, nil
+	}
+
+	if state, ok := a.Config().PluginSettings.PluginStates[plugin.Manifest.Id]; ok && state.Enable {
+		newVersion, err := isNewVersion(plugin.Manifest, avaliablePlugins)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to parse plugin version %s for prepackaged plugin %s", plugin.Manifest.Version, plugin.Manifest.Id)
+		}
+
+		if newVersion {
+			signatures, err := a.readSignaturesFromPath(pluginPaths)
+			if err != nil {
+				return nil, errors.Wrapf(err, "Failed to open plugin signatures for plugin %s", pluginPaths.pluginId)
+			}
+			fileReader.Seek(0, 0)
+			if err := a.VerifyPluginWithSignatures(fileReader, signatures); err != nil {
+				return nil, errors.Wrapf(err, "Failed to verify prepackaged plugin %s", pluginPaths.pluginId)
+			}
+			if _, err := a.installExtractedPlugin(plugin.Manifest, pluginDir, true); err != nil {
+				return nil, errors.Wrapf(err, "Failed to install extracted prepackaged plugin %s", pluginPaths.pluginId)
+			}
+		}
 	}
 
-	return plug, nil
+	return plugin, nil
+}
+
+// isNewVersion checks if plugin manifest has newer version of the plugin.
+func isNewVersion(manifest *model.Manifest, avaliablePlugins []*model.BundleInfo) (bool, error) {
+	for _, plugin := range avaliablePlugins {
+		if plugin.Manifest.Id == manifest.Id {
+			curVersion := semver.MustParse(plugin.Manifest.Version)
+			newVersion, err := semver.Parse(manifest.Version)
```

Comment:
We `MustParse` and `Parse` in this method -- propose we do one or the other here to avoid "special" knowledge about which one can be trusted to already be right.